### PR TITLE
Markdown example: four CommonMark quick wins

### PR DESCRIPTION
**Pull request by egonSchiele**
Closes four small CommonMark gaps in the markdown example, one phase per commit. Each commit is its own red→green→commit TDD cycle.

## Changes

| # | Feature | Commit |
|---|---|---|
| 1 | ATX heading level capped at 6 (7+ `#` falls through to a paragraph) | 1a8392d |
| 2 | Optional trailing `#` run stripped from ATX headings | c6e246a |
| 3 | Optional `"title"` / `'title'` on inline links and images | dbd6489 |
| 4 | Multi-backtick code spans with proper space stripping | e5d1585 |
| 5 | Soft-wrapped paragraphs with `inline-soft-break` | 1659d4f |
| 6 | README + CHANGELOG | 61e6b10 |

## AST changes (additive only)
- `InlineMarkdown` union gains `InlineSoftBreak = { type: "inline-soft-break" }`.
- `InlineLink` and `Image` gain optional `title?: string`.

## Design notes
- **ATX heading marker** is now `or(...[6,5,4,3,2,1].map(n => seqR(exactly(n, char("#")), not(char("#")))))` — widest first, with a trailing `not(char("#"))` to reject 7+ runs.
- **Trailing `#` stripping** captures the body via `many1Till(or(\n, trailingHashRun))` then re-parses it as inline markdown (same shape as the existing setext heading parser).
- **Link/image titles** share a new `urlToken` (whitespace- and `)`-terminated) and `titleClause` (leading-space-separated `"..."`/`'...'`) — the URL no longer slurps via `iManyTillStr(")")` so the title is unambiguous.
- **Code spans** use `count(char(\"`\"))` for the opening run, a body of `noneOf(\"`\")` atoms or tick-runs whose length is *not* N (`not(tickRunOf(n))` lookahead), then a closer of exactly N ticks not followed by another ```. The opener count threads into the closer via a small wrapper; everything else stays combinator-shaped.
- **Soft breaks** export a `softBreakParser` (single `\n` with `not(\n)` lookahead) that is *only* used inside `paragraphInline` — adding it to `inlineMarkdownParser`'s top-level `or` would have changed every other block's behaviour. Hard breaks (`"  \n"` / `"\\\n"`) still win because they're matched earlier in the same `or`.

## Validation
- `npx vitest run` → 396/396 passing (+20 new tests).
- `npm run test:tsc` clean modulo the two pre-existing errors on `main` (`regex` export, `vitest.globals` module).

### CODE REVIEW

File: tests/examples/markdown/inline.ts [hunk 1]
```diff
@@ -84,27 +89,95 @@ export const inlineItalicParser: Parser<InlineItalic> = map(
   ({ content }) => ({ type: "inline-italic" as const, content: content as InlineMarkdown[] })
 );
 
+/* URL + optional title used by both inline-link and inline-image parsers.
+ * `urlToken` is whitespace- and `)`-terminated. `titleClause` is an optional
+ * leading-space-separated `"..."` or `'...'`. Both are pure combinator-based
+ * so the link/image parsers can share them without special casing. */
+const urlToken: Parser<string> = many1WithJoin(noneOf(" \t\n)"));
+
+const dqTitle: Parser<string> = map(
```
Comment (egonSchiele) [inline tests/examples/markdown/inline.ts]: use better var names, what is `dq`? Also `map` is unnecessary here

Comment (egonSchiele) [reply tests/examples/markdown/inline.ts]: You're right on both counts — `dq`/`sq` were opaque and the per-quote-style `map` wrappers were unnecessary. Collapsed both helpers into a single inline `captureCaptures(or(seqC(\"\\\"\", capture(..., \"title\"), \"\\\"\"), seqC(\"'\", capture(..., \"title\"), \"'\")))` block. `captureCaptures` lifts the per-arm `title` capture into the outer `titleClause`'s captures so the final `map` only has to destructure once. Fixed in ecf0f1c.

File: tests/examples/markdown/inline.ts [hunk 2]
```diff
@@ -84,27 +89,95 @@ export const inlineItalicParser: Parser<InlineItalic> = map(
   ({ content }) => ({ type: "inline-italic" as const, content: content as InlineMarkdown[] })
 );
 
+/* URL + optional title used by both inline-link and inline-image parsers.
+ * `urlToken` is whitespace- and `)`-terminated. `titleClause` is an optional
+ * leading-space-separated `"..."` or `'...'`. Both are pure combinator-based
+ * so the link/image parsers can share them without special casing. */
+const urlToken: Parser<string> = many1WithJoin(noneOf(" \t\n)"));
```
[the remark this answers is not in the document]
Comment (egonSchiele) [reply tests/examples/markdown/inline.ts]: Correct — the old `iManyTillStr(\")\")` allowed empty URLs but `many1WithJoin(noneOf(...))` regressed that. Switched `urlToken` to `manyWithJoin(noneOf(\" \\t\\n)\"))` so `[a]()` and `alt` parse with `url: \"\"`; the title disambiguation still works because `urlToken` won't consume the leading space before a title and `titleClause` requires one. Added explicit tests for both. Fixed in ecf0f1c.

File: tests/examples/markdown/inline.ts [hunk 3]
```diff
@@ -84,27 +89,95 @@ export const inlineItalicParser: Parser<InlineItalic> = map(
...
-    content: content as InlineMarkdown[],
-    url,
-  })
+  ({ content, url, title }) => {
+    const link: InlineLink = {
+      type: "inline-link",
+      content: content as InlineMarkdown[],
+      url,
+    };
+    if (title) link.title = title;
```
[the remark this answers is not in the document]
Comment (egonSchiele) [reply tests/examples/markdown/inline.ts]: Right — `if (title)` was a truthiness bug that silently dropped an explicit empty title (`[a](u \"\")`). `optional(titleClause)` returns `string | null`, so swapped to `if (title != null)` which preserves `\"\"` while still skipping the no-title-clause path. Added a regression test. Fixed in ecf0f1c.

File: tests/examples/markdown/inline.ts [hunk 4]
```diff
@@ -240,15 +313,23 @@ export const inlineRefImageParser: Parser<InlineRefImage> = map(
...
+    str("!["),
+    capture(iManyTillStr("]("), "alt"),
+    str("]("),
+    capture(urlToken, "url"),
+    capture(optional(titleClause), "title"),
+    char(")")
+  ),
+  ({ alt, url, title }) => {
+    const img: Image = { type: "image", alt, url };
+    if (title) img.title = title;
```
[the remark this answers is not in the document]
Comment (egonSchiele) [reply tests/examples/markdown/inline.ts]: Same root cause and same fix as the `InlineLink` thread — swapped `if (title)` to `if (title != null)` in `imageParser` so `alt` preserves the empty title. Added a matching regression test. Fixed in ecf0f1c.

File: tests/examples/markdown/blocks.ts [hunk 5]
```diff
@@ -364,8 +401,15 @@ export const horizontalRuleParser: Parser<HorizontalRule> = or(
...
+/* A paragraph node: an inline node OR a soft line break (single `\n` that
+ * isn't the start of a blank line). Hard breaks ("  \n" / "\\\n") win over
+ * soft breaks because they're matched earlier inside `inlineMarkdownParser`'s
+ * `or`. */
 const paragraphInline: Parser<InlineMarkdown> = map(
-  seqC(not(blankLine), capture(inlineMarkdownParser, "node")),
+  seqC(
+    not(blankLine),
+    capture(or(softBreakParser, inlineMarkdownParser), "node")
+  ),
```
[the remark this answers is not in the document]
Comment (egonSchiele) [reply tests/examples/markdown/blocks.ts:439]: Good catch — a real bug. `para\\n# heading` was greedily absorbed as one soft-wrapped paragraph because nothing rejected the soft break when the next line started a block.

Added a `blockInterrupt` parser in `blocks.ts` that matches the start of an ATX heading (`atxMarker` + space), block quote (`>`), fenced code (`\\\`\\\`\\\``), HR, list marker, table row, or HTML block opener; then wrapped `softBreakParser` in `seqR(softBreakParser, not(blockInterrupt))` to form `paragraphSoftBreak`. `seq`'s atomic rollback restores the `\\n` when the lookahead fails, so the paragraph terminates and the next block parser picks up cleanly.

Setext is intentionally **not** in `blockInterrupt` — its underline resolution depends on the heading text being read by the paragraph parser first, and `setextHeadingParser` runs before `paragraphParser` in the top-level dispatch.

Added 8 `it.each` regression tests covering each block opener type. Fixed in be1404b.

**Comment by egonSchiele** [comment]
Addressed all five review comments across two commits:

**ecf0f1c — `fix(markdown): link/image edge cases`**
- `urlToken`: `many1WithJoin` → `manyWithJoin` so empty destinations (`[a]()`, `alt`) parse.
- `InlineLink` and `Image` title assignment: `if (title)` → `if (title != null)` so explicitly empty titles are preserved.
- `titleClause`: collapsed the `dq`/`sq` helpers + per-quote-style `map`s into a single inline `captureCaptures(or(seqC..., seqC...))` block.

**be1404b — `fix(markdown): interrupt soft-wrapped paragraph at block openers`**
- New `blockInterrupt` parser (ATX heading, blockquote, fenced code, HR, list marker, table row, HTML block opener) used as a `not(...)` lookahead on the consumed `\\n` of `paragraphSoftBreak`. Setext intentionally excluded — handled by its earlier dispatch position.

Verification: `npx vitest run` → 408/408 passing (+12 new regression tests); `npm run test:tsc` clean modulo the two pre-existing errors on `main`.
